packages/apollo-language-server/src/utilities/graphql.ts

```diff
@@ -87,31 +93,180 @@ export function removeDirectives(ast: ASTNode, directiveNames: string[]) {
   });
 }
 
-// remove fields where a given directive is found
-export function removeDirectiveAnnotatedFields(
-  ast: ASTNode,
+/**
+ * Recursively remove orphaned fragment definitions that have their names included in
+ * `fragmentNamesEligibleForRemoval`
+ *
+ * We expclitily require the fragments to be listed in `fragmentNamesEligibleForRemoval` so we only strip
+ * fragments that were orphaned by an operation, not fragments that started as oprhans
+ */
+function removeOrphanedFragmentDefinitions<AST extends ASTNode>(
+  ast: AST,
+  fragmentNamesEligibleForRemoval: Set<string>
+): AST {
+  /**
+   * Flag to keep track of removing any fragments
+   */
+  let anyFragmentsRemoved = false;
+
+  // Aquire names of all fragment spreads
+  const fragmentSpreadNodeNames = new Set<string>();
+  visit(ast, {
+    FragmentSpread(node) {
+      fragmentSpreadNodeNames.add(node.name.value);
+    }
+  });
+
+  // Strip unused fragment definitions. Flag if we've removed any so we know if we need to continue
+  // recursively checking.
+  ast = visit(ast, {
+    FragmentDefinition: {
+      enter(node) {
+        if (
+          fragmentNamesEligibleForRemoval.has(node.name.value) &&
+          !fragmentSpreadNodeNames.has(node.name.value)
+        ) {
+          // This definition is not used, remove it.
+          anyFragmentsRemoved = true;
+          return null;
+        }
+
+        return undefined;
+      }
+    },
+    FragmentSpread: {
+      leave(node) {
+        if (!fragmentSpreadNodeNames.has(node.name.value)) {
+          // This definition is not used, remove it.
+          anyFragmentsRemoved = true;
+          return null;
+        }
+
+        return undefined;
+      }
+    }
+  });
+
+  if (anyFragmentsRemoved) {
+    // We've removed fragments and might have orphaned more fragments, so recursively try to remove more
+    // orphaned fragments.
+    return removeOrphanedFragmentDefinitions(
```

Comment:
So this is handling the case of:

```graphql
fragment Two on Node {
  id
}
fragment One on Query {
  hero {
    ...Two @client 
  }
}
{ ...One }
```
Where the visitor goes past the `Two` fragment first since it isn't know that it needs to be trimmed. After we run through and see that `One` has the only usage of `Two` and it is client only, it has now become orphaned.

If I understand this correctly, I think we should explain the recursion in a comment with a reference query like above